graders/python/python_autograder/run.sh

```diff
@@ -0,0 +1,71 @@
+#! /bin/bash
+
+##########################
+# INIT
+##########################
+
+# the autograder directory
+AG_DIR='/python_autograder'
+
+# the parent directory containing everything about this grading job
+export JOB_DIR='/grade'
+# the job subdirectories
+STUDENT_DIR=$JOB_DIR'/student'
+TEST_DIR=$JOB_DIR'/tests'
+OUT_DIR=$JOB_DIR'/results'
+
+# where we will copy everything
+export MERGE_DIR=$JOB_DIR'/run'
+
+# now set up the stuff so that our run.sh can work
+mkdir $MERGE_DIR
+mkdir $OUT_DIR
+
+mv $STUDENT_DIR/* $MERGE_DIR
+mv $AG_DIR/* $MERGE_DIR
+mv $TEST_DIR/* $MERGE_DIR
+
+# user does not need a copy of this script
+rm -f "$MERGE_DIR/run.sh"
+
+# we need this to include code as python modules
+touch $MERGE_DIR/__init__.py
+
+# Do not allow ag user to modify, rename, or delete any existing files
+chmod -R 755 "$MERGE_DIR"
+chmod 1777 "$MERGE_DIR"
+
+# Create directory without sticky bit for deletable files
+export FILENAMES_DIR=$MERGE_DIR'/filenames'
+mkdir $FILENAMES_DIR
+chmod 777 $FILENAMES_DIR
+mv ans.py setup_code.py test.py $FILENAMES_DIR
+
+##########################
+# RUN
+##########################
+
+echo "[run] starting autograder"
+
+# randomly generate the name of the results file, so that someone can't guess and write to it
+SECRET_NAME=$MERGE_DIR/`uuidgen`
+
+# run the autograder as a limited user called ag
+su -c "python3 $MERGE_DIR/pl_main.py $SECRET_NAME" ag
```

Comment:
Command-line arguments are globally visible (e.g., student code can run `ps` to see the arguments of commands run as root). The two standard methods for conveying secret information are to pass it via STDIN, or to use a temporary file that is deleted after reading (like the code used to do). I'd be happy to go back to the secrets-file-that-will-be-deleted method. I think my earlier comment about this was just to explain in the code why we are doing it (specifically, to avoid passing secret command-line arguments).